930 o share icons (#951)
Financial-Times/origami: components/o-share/MIGRATION.md

# Migration guide

## Migrating from v8 to v9

> Version 9 has markup changes and component API got updated as well. New version relies less on javaScript and improves accessibility on high contrast mode. The most important updates are:
> 
> 1. The icons are now inlined inside the templates, since this was our only known option to [fix accessibility issue](https://github.com/Financial-Times/origami/issues/930) on high contrast mode devices.
> 2. Share link MUST be written in full markup instead of using js to automatically generate share.
> 3. New TSX template requires social icons to be passed as children.

Reviewer: This list describes what has changed and why which can be helpful context but could also distract from the steps a user needs to take to migrate. Could these be updated to be a kind of summary of steps which are detailed below?

```suggestion
o-share v9 incudes a number of changes to improve accessibility and the flexibility of its TSX templates. To migrate follow these steps and see below for more details:

1. If your project uses o-share TSX templates, migrate to the new template interface.
2. o-share no longer provides client side JavaScript to generate markup. If your project does not use o-share TSX templates ensure you output o-share markup in full. 
```

> New markup can be copied from Origami [registry](https://registry.origami.ft.com/components/o-share) or Build service users can use [storyBook demos](http://localhost:6969/?path=/story/components-o-share--horizontal) to generate markup based on storybook controls, that can be copied from `HTML` tab.

Reviewer: Correcting a link here:
```suggestion
New markup can be copied from the [Origami registry](https://registry.origami.ft.com/components/o-share) or [StoryBook](https://origami.ft.com/storybook/) via the`HTML` tab.
```

### Inlined SVG icons

> Version 8 was utilising `:before` selector and icon url was dynamically set using scss. This approach was not working on high contrast mode devices. We have decided to inline SVG icons instead. This is a breaking change since we have removed `:before` pseudo element and in version 9 we use following structure for `o-share` items:

Reviewer: This  describes what has changed and not what a user needs to do. Do they need to know we used to use the `:before` selector?

Reviewer 2: I agree we can simplify this to just the information the user needs. @notlee Is there a decision log somewhere which we can reference should the user wish to read up on our decision to use SVGs?

Reviewer: Do you have a formal process/doc in mind regarding how a decision log should work? We don't have anything quite like that, but do document decisions in a few places depending on what they are.

I think for this I'd propose putting this information in the commit message and PR message of the change. It could also make an interesting blog post but in addition to, as a blog post isn't very discoverable when working on/with o-share. For significant or cross-component changes we'd usually have an issue with label `proposal` which would log a decision and the conversation around that.

Reviewer 2: In the past we've logged them on a `.md` in a repository which served as a hub for a given team (for us it would be something like `origami-team-home`. Notion is also a good option that can give some additional features such as filtering based on certain properties.

I'm thinking because the team is quite small, we might not need a hugely formal DL as we should be able to share knowledge and be on the same page easier than a larger team. Though DLs have use for external/in-source contributors.

For now we can log this in the commit and PR. Would this be the release please PR?

Reviewer: This commit and PR, it would be easier for a future person to `git blame` their way here

```html
<li class="o-share__action">
   <a
    class="o-share__icon o-share__icon--{{className}}"
    href="{{link}}"
    rel="noopener"
   >
    <div class="o-share__icon__image">
    <!-- SVG code -->
    </div>
    <span class="o-share__text">
     Share {{title}} on {{name}} (opens a new window)
    </span>
   </a>
</li>
```

### Deprecate autogenerate share links

> Before version 9 users could automatically generate share links with following markup. But this requires JavaScript and from our lates DAC audit we determined that this approach could be deprecated.

Reviewer: This isn't quite true. We removed this feature as there were multiple ways to use the component which is difficult to maintain. We felt that tsx templates provide a reasonable alternative for users who do not want to manage markup. But probably we don't need quite so much detail in a migration guide away 😄 That might be better for the commit description and changelog

```html
<div data-o-component="o-share"
 class="o-share"
 data-o-share-links="{{links}}"
 data-o-share-url="{{url}}"
 data-o-share-title="{{title}}"
 data-o-share-titleExtra="{{titleExtra}}"
 data-o-share-summary="{{summary}}"
 data-o-share-relatedTwitterAccounts="{{relatedTwitterAccounts}}"
 data-o-share-location="{{locationOfShareComponent}}">
</div>
```

Instead use full markup of the component:

> ```html

Reviewer: This is a helpful example 👍

<!-- see the registry demos for full markup -->
<div data-o-component="o-share" class="o-share">
 <ul>
  <!-- a share to twitter action example -->
  <!-- href tag is not shown, see the registry demos for full markup  -->
  <li class="o-share__action">
   <a class="o-share__icon o-share__icon--twitter"
    href="#twitter-link-here"
    rel="noopener">
    <span class="o-share__text">Twitter</span>
   </a>
  </li>
  <!-- more o-share actions -->
 </ul>
</div>
```

### TSX template

`<Share />` component now must have each icon as children. To render icons you must use `<ShareIcon>` component and provide correct props:

```jsx
<Share {...shareProps}>
   <ShareIcon icon="twitter" urlProps={shareIconProps} />
   <ShareIcon
    icon="share"
    showLabel={true}
    label="Share"
    customAction="#"
    urlProps={shareIconProps}
   />
  </Share>
```

> This example should render twitter social share icon and custom share button. More details about the props that you must pass down to each compnent can be found in our [readme](README.md#tsx-template).

Reviewer: We removed custom share buttons for now
